# Sample Code for making a basic image classifier from scratch


### Instructions

This notebook is only to get you started with a code sample of a simple classifier built from scratch. The data on which models in this notebook are train are not provided. Hence, the output is given. You need not run the cells to see the output. If you choose to do so with your own training and testing dataset, you might not get the same accuracy.

DISCLAIMER: These models are not the best classifiers, they may perform very moderately on other data, the goal is to get new students started with the basic code required to build classifiers from scratch. Also, many more modifications may be implemented in terms of model architecture, data augmentation and other techniques.

### Simple Neural Network

For those who are new to this field, the first step is making a simple Densly connected neural network and training it with the data extracted from google images.

This is shown in the following cell, you can change the layers/architecture etc. to see if the accuracy increases.

In [8]:
# Try a simple Neural Network

import os
import cv2
import tensorflow as tf
from tensorflow.keras import layers, models
from sklearn.model_selection import train_test_split
import numpy as np

# Load and preprocess the data
def load_and_preprocess_data(cat_path, dog_path, target_size=(150, 150)):
    cat_images = [cv2.resize(cv2.imread(os.path.join(cat_path, f)), target_size) for f in os.listdir(cat_path)]
    dog_images = [cv2.resize(cv2.imread(os.path.join(dog_path, f)), target_size) for f in os.listdir(dog_path)]

    # Convert images to numpy arrays
    cat_data = np.array(cat_images)
    dog_data = np.array(dog_images)

    # Create labels (0 for cats, 1 for dogs)
    cat_labels = np.zeros(len(cat_data))
    dog_labels = np.ones(len(dog_data))

    # Concatenate data and labels
    data = np.concatenate((cat_data, dog_data))
    labels = np.concatenate((cat_labels, dog_labels))

    # Normalize pixel values to the range [0, 1]
    data = data / 255.0

    return data, labels

# Set paths to your cat and dog image folders
cat_path = 'Dataset/elephant'
dog_path = 'Dataset/horse'

# Load and preprocess data
data, labels = load_and_preprocess_data(cat_path, dog_path)

# Split the data into training and testing sets
train_data, test_data, train_labels, test_labels = train_test_split(
    data, labels, test_size=0.2, random_state=42
)

target_size = [150,150]

# Build the model
model = models.Sequential()
model.add(layers.Flatten(input_shape=(target_size[0], target_size[1], 3)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(20, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(train_data, train_labels, epochs=20, batch_size=32, validation_split=0.2)

# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(test_data, test_labels)
print(f"Test Accuracy: {test_accuracy}")


Epoch 1/20
8/8 [==============================] - 1s 66ms/step - loss: 6.2706 - accuracy: 0.4727 - val_loss: 3.1323 - val_accuracy: 0.4844
Epoch 2/20
8/8 [==============================] - 0s 46ms/step - loss: 2.2030 - accuracy: 0.5195 - val_loss: 1.5041 - val_accuracy: 0.5312
Epoch 3/20
8/8 [==============================] - 0s 50ms/step - loss: 1.6055 - accuracy: 0.4883 - val_loss: 1.0742 - val_accuracy: 0.5625
Epoch 4/20
8/8 [==============================] - 0s 48ms/step - loss: 1.1651 - accuracy: 0.5117 - val_loss: 1.7158 - val_accuracy: 0.4844
Epoch 5/20
8/8 [==============================] - 0s 46ms/step - loss: 1.3062 - accuracy: 0.5664 - val_loss: 1.7494 - val_accuracy: 0.4844
Epoch 6/20
8/8 [==============================] - 0s 44ms/step - loss: 1.0671 - accuracy: 0.5547 - val_loss: 1.0123 - val_accuracy: 0.5625
Epoch 7/20
8/8 [==============================] - 0s 44ms/step - loss: 0.8387 - accuracy: 0.5820 - val_loss: 0.6953 - val_accuracy: 0.6562
Epoch 8/20
8/8 [===========

#### Output:

Epoch 1/20

8/8 [==============================] - 1s 66ms/step - loss: 6.2706 - accuracy: 0.4727 - val_loss: 3.1323 - val_accuracy: 0.4844

Epoch 2/20

8/8 [==============================] - 0s 46ms/step - loss: 2.2030 - accuracy: 0.5195 - val_loss: 1.5041 - val_accuracy: 0.5312

Epoch 3/20

8/8 [==============================] - 0s 50ms/step - loss: 1.6055 - accuracy: 0.4883 - val_loss: 1.0742 - val_accuracy: 0.5625

Epoch 4/20

8/8 [==============================] - 0s 48ms/step - loss: 1.1651 - accuracy: 0.5117 - val_loss: 1.7158 - val_accuracy: 0.4844

Epoch 5/20

8/8 [==============================] - 0s 46ms/step - loss: 1.3062 - accuracy: 0.5664 - val_loss: 1.7494 - val_accuracy: 0.4844

Epoch 6/20

8/8 [==============================] - 0s 44ms/step - loss: 1.0671 - accuracy: 0.5547 - val_loss: 1.0123 - val_accuracy: 0.5625

Epoch 7/20

8/8 [==============================] - 0s 44ms/step - loss: 0.8387 - accuracy: 0.5820 - val_loss: 0.6953 - val_accuracy: 0.6562

Epoch 8/20

8/8 [==============================] - 0s 45ms/step - loss: 0.8533 - accuracy: 0.6094 - val_loss: 1.0579 - val_accuracy: 0.4844

Epoch 9/20

8/8 [==============================] - 0s 45ms/step - loss: 0.9174 - accuracy: 0.6602 - val_loss: 0.9649 - val_accuracy: 0.4844

Epoch 10/20

8/8 [==============================] - 0s 46ms/step - loss: 0.7389 - accuracy: 0.6523 - val_loss: 0.6622 - val_accuracy: 0.6719

...

Epoch 20/20

8/8 [==============================] - 0s 43ms/step - loss: 0.3647 - accuracy: 0.8281 - val_loss: 0.7517 - val_accuracy: 0.6094

3/3 [==============================] - 0s 5ms/step - loss: 0.9466 - accuracy: 0.5250

Test Accuracy: 0.5249999761581421

### CNN (Convolutional Neural Networks)

Now that you've tried using a simple densely connected Neural Network, you should use CNNs (Convolutional Neural Networks), these are widely used as the base of several simple image classifiers.

In [10]:
import os
import cv2
import tensorflow as tf
from tensorflow.keras import layers, models
from sklearn.model_selection import train_test_split
import numpy as np

# Load and preprocess the data
def load_and_preprocess_data(cat_path, dog_path, target_size=(150, 150)):
    cat_images = [cv2.resize(cv2.imread(os.path.join(cat_path, f)), target_size) for f in os.listdir(cat_path)]
    dog_images = [cv2.resize(cv2.imread(os.path.join(dog_path, f)), target_size) for f in os.listdir(dog_path)]

    # Convert images to numpy arrays
    cat_data = np.array(cat_images)
    dog_data = np.array(dog_images)

    # Create labels (0 for cats, 1 for dogs)
    cat_labels = np.zeros(len(cat_data))
    dog_labels = np.ones(len(dog_data))

    # Concatenate data and labels
    data = np.concatenate((cat_data, dog_data))
    labels = np.concatenate((cat_labels, dog_labels))

    # Normalize pixel values to the range [0, 1]
    data = data / 255.0

    return data, labels

# Set paths to your cat and dog image folders
cat_path = 'Dataset/elephant'
dog_path = 'Dataset/horse'

# Load and preprocess data
data, labels = load_and_preprocess_data(cat_path, dog_path)

# Split the data into training and testing sets
train_data, test_data, train_labels, test_labels = train_test_split(
    data, labels, test_size=0.2, random_state=42
)

target_size = [150,150]

# Build a CNN model
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(target_size[0], target_size[1], 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(train_data, train_labels, epochs=30, batch_size=32, validation_split=0.2)

# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(test_data, test_labels)
print(f"Test Accuracy: {test_accuracy}")


Epoch 1/30
8/8 [==============================] - 3s 247ms/step - loss: 0.8998 - accuracy: 0.5000 - val_loss: 0.7213 - val_accuracy: 0.4844
Epoch 2/30
8/8 [==============================] - 2s 226ms/step - loss: 0.6930 - accuracy: 0.5586 - val_loss: 0.6915 - val_accuracy: 0.4844
Epoch 3/30
8/8 [==============================] - 2s 224ms/step - loss: 0.6928 - accuracy: 0.5547 - val_loss: 0.6855 - val_accuracy: 0.5469
Epoch 4/30
8/8 [==============================] - 2s 219ms/step - loss: 0.6572 - accuracy: 0.5977 - val_loss: 0.7093 - val_accuracy: 0.4688
Epoch 5/30
8/8 [==============================] - 2s 223ms/step - loss: 0.5845 - accuracy: 0.6797 - val_loss: 0.7011 - val_accuracy: 0.5469
Epoch 6/30
8/8 [==============================] - 2s 224ms/step - loss: 0.5682 - accuracy: 0.6641 - val_loss: 0.5912 - val_accuracy: 0.6406
Epoch 7/30
8/8 [==============================] - 2s 222ms/step - loss: 0.5125 - accuracy: 0.7500 - val_loss: 0.6445 - val_accuracy: 0.6719
Epoch 8/30
8/8 [====

#### Output

Epoch 1/30

8/8 [==============================] - 3s 247ms/step - loss: 0.8998 - accuracy: 0.5000 - val_loss: 0.7213 - val_accuracy: 0.4844

Epoch 2/30

8/8 [==============================] - 2s 226ms/step - loss: 0.6930 - accuracy: 0.5586 - val_loss: 0.6915 - val_accuracy: 0.4844

Epoch 3/30

8/8 [==============================] - 2s 224ms/step - loss: 0.6928 - accuracy: 0.5547 - val_loss: 0.6855 - val_accuracy: 0.5469

...

Epoch 30/30

8/8 [==============================] - 2s 255ms/step - loss: 5.2710e-04 - accuracy: 1.0000 - val_loss: 1.1712 - val_accuracy: 0.7969

3/3 [==============================] - 0s 53ms/step - loss: 2.1892 - accuracy: 0.7125

Test Accuracy: 0.7124999761581421

### Data Augmentation

To increase the accuracy further, we must augment the data, ie; change the training data (such that the original subject of use is still meaningful). This involves zooming in/out, horizontal flipping, rotation etc. This helps the model learn modifications that it might encounter ahead.

In [12]:
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Split the data into training and validation sets
train_data, val_data, train_labels, val_labels = train_test_split(
    data, labels, test_size=0.2, random_state=42
)

# Augment the training data
datagen = ImageDataGenerator(
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest'
)

# Fit the data generator on the training set
datagen.fit(train_data)

# Train the model with augmented data
model.fit(
    datagen.flow(train_data, train_labels, batch_size=32),
    epochs=30,
    validation_data=(val_data, val_labels)
)


Epoch 1/30
10/10 [==============================] - 3s 266ms/step - loss: 1.2988 - accuracy: 0.5969 - val_loss: 0.7283 - val_accuracy: 0.6000
Epoch 2/30
10/10 [==============================] - 3s 262ms/step - loss: 0.6696 - accuracy: 0.5625 - val_loss: 0.7129 - val_accuracy: 0.4625
Epoch 3/30
10/10 [==============================] - 3s 270ms/step - loss: 0.6564 - accuracy: 0.5406 - val_loss: 0.7108 - val_accuracy: 0.4750
Epoch 4/30
10/10 [==============================] - 3s 259ms/step - loss: 0.6581 - accuracy: 0.5594 - val_loss: 0.6827 - val_accuracy: 0.5750
Epoch 5/30
10/10 [==============================] - 3s 262ms/step - loss: 0.6501 - accuracy: 0.6531 - val_loss: 0.6808 - val_accuracy: 0.5750
Epoch 6/30
10/10 [==============================] - 3s 263ms/step - loss: 0.6185 - accuracy: 0.6438 - val_loss: 0.7264 - val_accuracy: 0.6000
Epoch 7/30
10/10 [==============================] - 3s 262ms/step - loss: 0.6016 - accuracy: 0.6562 - val_loss: 0.6720 - val_accuracy: 0.6250
Epoch 

In [13]:
# Evaluate the model on the test set
test_loss, test_accuracy = model.evaluate(test_data, test_labels)
print(f"Test Accuracy: {test_accuracy}")

3/3 [==============================] - 0s 45ms/step - loss: 0.5904 - accuracy: 0.7625
Test Accuracy: 0.762499988079071


#### Output
3/3 [==============================] - 0s 45ms/step - loss: 0.5904 - accuracy: 0.7625

Test Accuracy: 0.762499988079071

### Further steps

A basic model with acceptable accuracy and effort is enough for midterm evaluations.

You are encouraged to look into open source libraries for more complicated models to get a good postition on the Kaggle leaderboard!

### Happy coding!